aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -13,80 +13,132 @@ namespace native {
-  const Tensor& in_tensor,
-  Tensor& out_tensor
-){
-  const int64_t numel = in_tensor.numel();
-  const scalar_t* in_tensor_d = in_tensor.data_ptr<scalar_t>();
-  scalar_t* out_tensor_d = out_tensor.data_ptr<scalar_t>();
-  auto sizes_v = in_tensor.sizes().vec();
-  auto strides_v = in_tensor.strides().vec();
-
-  at::parallel_for(0, numel, 1000, [&](int64_t start, int64_t end) {
-    for (auto i = start; i < end; i++) {
-      int64_t cur_indices = i;
-      int64_t rem = 0;
-      int64_t dst_offset = 0;
-
-      for (int64_t d = 0; d < total_dims; d++) {
-        int64_t temp = cur_indices;
-        cur_indices = cur_indices / stride_contiguous_v[d];
-        rem = temp - cur_indices * stride_contiguous_v[d];
-        dst_offset += flip_dims_b[d] ? (sizes_v[d] - 1 - cur_indices) * strides_v[d] : cur_indices * strides_v[d];
-        cur_indices = rem;
-      }
-      out_tensor_d[i] = in_tensor_d[dst_offset];
+struct Indexer {
+  Indexer(int64_t num_indexers, char** indexers, const int64_t* indexer_strides)
+    : num_indexers(num_indexers)
+    , indexers(indexers)
+    , indexer_strides(indexer_strides) {
+
+  }
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = *(int64_t*)&indexers[0][idx * indexer_strides[0]];
+    for (int j = 1; j < num_indexers; j++) {
+      offset += *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+      // offset += value;
     }
-  });
+    return offset;
+  }
+};
+
+Tensor build_index(int64_t num_dims, int64_t flip_dim, int64_t dim_size) {
+  auto new_shape = std::vector<int64_t>(num_dims, 1);
+  new_shape[flip_dim] = dim_size;
+
+  TensorOptions tensor_options =
+    TensorOptions(c10::kLong).
+    device(c10::kCPU);
+
+  return at::empty(new_shape, tensor_options);
 }
 
-Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
-  auto in_tensor = self;
-  const int64_t total_dims = in_tensor.dim();
-  auto flip_dims_b = at::dim_list_to_bitset(dims, total_dims);
-  Tensor out_tensor = at::empty_like(in_tensor, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-
-  // create contiguous strides for input tensor
-  auto stride_contiguous_v = std::vector<int64_t>(total_dims);
-  for (int64_t i = total_dims - 1; i >= 0; i--) {
-    if (i == total_dims - 1) {
-      stride_contiguous_v[i] = 1;
-    } else {
-      stride_contiguous_v[i] = std::max<int64_t>(in_tensor.size(i + 1), 1) * stride_contiguous_v[i + 1];
+
+std::vector<Tensor> build_indices_loop(Tensor input, IntArrayRef flip_dims) {
+  std::vector<Tensor> indices;
+  int64_t element_size_bytes = input.element_size();
+  for(auto dim: flip_dims) {
+    auto dim_size = input.size(dim);
+    auto index = build_index(input.ndimension(), dim, dim_size);
+    auto stride = input.stride(dim);
+    auto input_index_ptr = index.data_ptr<int64_t>();
+
+    for(int64_t i = 0; i < dim_size; i++) {
+      input_index_ptr[i] = static_cast<int64_t>(dim_size - i - 1) * stride * element_size_bytes;
```

Comment:
I think you could move this logic to `build_index`, as of now `build_index` only returns a non-filled tensor, which is a bit weird.

And the signature of `build_index` could be `Tensor input, int64_t flip_dim`, and you can extract the other information from `input` itself.